Refactor to work with `credo` JSON output

larkinscott opened the pull request:
Borrowing some code from: rrrene/credo#592

The maintainer of `credo` brought up a good point on the PR referenced
above, which is that this plugin can work with the output from credo's
JSON output instead of trying to merge a formatter upstream. This is how
most of our plugins work, actually.

This refactor allows us to pull down the latest version of `credo`
instead of maintaining our own version (the previous version was using a
fork of `credo`).

If this is ok, we should cut over to using this version of `codeclimate-credo` so we can take advantage of the latest `credo` versions.

larkinscott committed:
Bump `codeclimate-credo`

The `codeclimate-credo` plugin is currently using an older version of
`credo` and we'd like to update it to the most recent version.

This change pulls a forked version of `credo`. The current
version of `codeclimate-credo` in production is using a fork of `credo`
from the maintainer with tweaks that enable `credo` to work with Code
Climate's spec. Ideally, we'd like to get those changes merged in
upstream, but just preparing this in case we aren't able to.

larkinscott committed:
Refactor

Borrowing some code from: https://github.com/rrrene/credo/pull/592

The maintainer of `credo` brought up a good point on the PR referenced
above, which is that this plugin can work with the output from credo's
JSON output instead of trying to merge a formatter upstream. This is how
most of our plugins work, actually.

This refactor allows us to pull down the latest version of `credo`
instead of maintaining our own version (the previous version was using a
fork of `credo`).

filipesperandio commented on Dockerfile:
```diff
@@ -29,8 +29,7 @@ RUN cd bunt && MIX_ENV=prod mix deps.get --force
 
-RUN git clone https://github.com/fazibear/credo
-RUN cd credo && git checkout codeclimate
+RUN git clone https://github.com/rrrene/credo
```
We could probably use something like `--depth 1` to skip the history.

filipesperandio reviewed (approved):
I think this is a good thing!

wfleming commented on Makefile:
```diff
@@ -0,0 +1,7 @@
+
+image:
+	 docker build --rm -t $(IMAGE_NAME) .
+
```
:lipstick: blank trailing line can be removed

wfleming commented on lib/code_climate.ex:
```diff
@@ -5,7 +5,24 @@ defmodule CodeClimate do
+  @default_opts ~w[credo --format=json]
+
+  @issue_category %{
+    "Credo.Check.Design.DuplicatedCode" => ["Duplication"],
```
Might be worth extracting the category names to constants? e.g. `@category_duplication "Duplication"`. Up to you.

wfleming commented on lib/code_climate.ex:
```diff
@@ -43,7 +129,7 @@ defmodule CodeClimate do
   defp process_exclude_paths(opts, %{"exclude_paths" => exclude_paths}, path) do
```
Do we still expect plugins to process exclude paths? The CLI hasn't sent them down to plugins in like 4 years.

larkinscott replied:
Good catch - thanks!

wfleming reviewed (approved):
Small suggestions, not blocking. I also like Filipe's suggestion about the clone.

larkinscott committed:
Tweaks

- Make issue categories constants
- Remove exclude paths since we don't send those to the CLI and haven't
for quite a while
- Some formatting cleanup
